### PR TITLE
Python api: Various fixes in docs.

### PR DESCRIPTION
**Pull request by aniemetz**
Note: Fixes for kind documentation will be submitted separately.

### CODE REVIEW

File: src/api/python/cvc5.pxi [hunk 1]
```diff
@@ -99,7 +99,8 @@ cdef c_hash[c_Term] ctermhash = c_hash[c_Term]()
 cdef class Datatype:
     """
         A cvc5 datatype.
-        Wrapper class for :cpp:class:`cvc5::Datatype`.
+
+        Wrapper class for the C++ class :cpp:class`cvc5::Datatype`.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:103]:
```suggestion
        Wrapper class for the C++ class :cpp:class:`cvc5::Datatype`.
```

File: src/api/python/cvc5.pxi [hunk 2]
```diff
@@ -504,15 +568,16 @@ cdef class Op:
...
+            :return: The number of indices of this op.
         """
         return self.cop.getNumIndices()
 
     def __getitem__(self, i):
         """
-            Get the index at position i.
-            :param i: the position of the index to return
-            :return: the index at position i
+            Get the index at position ``i``.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:577]: `__getitem__` is a special member function and not documented by default. It must be added explicitly in `op.rst`, for example with `:special-members`.

Comment (aniemetz) [reply src/api/python/cvc5.pxi:577]: Right, do we want to add it?

File: src/api/python/cvc5.pxi [hunk 3]
```diff
@@ -966,13 +1077,15 @@ cdef class Solver:
...
         """
-        Supports the following arguments:
+            Create a term.
+
+            Supports the following arguments:
 
-        - ``Term mkTerm(Kind kind)``
-        - ``Term mkTerm(Kind kind, Op child1, List[Term] children)``
-        - ``Term mkTerm(Kind kind, List[Term] children)``
+            - ``Term mkTerm(Kind kind)``
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/api/python/cvc5.pxi:1084]: The second case is supported: https://github.com/cvc5/cvc5/blob/main/test/unit/api/python/test_solver.py#L795

File: src/api/python/cvc5.pxi [hunk 4]
```diff
@@ -1010,11 +1125,13 @@ cdef class Solver:
     @expand_list_arg(num_req_args=0)
     def mkOp(self, k, *args):
         """
-        Supports the following uses:
+            Create operator.
 
-        - ``Op mkOp(Kind kind)``
-        - ``Op mkOp(Kind kind, const string& arg)``
-        - ``Op mkOp(Kind kind, uint32_t arg0, ...)``
+            Supports the following arguments:
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1131]: Here, I think the `expand_list_arg` is not properly used: the argument should be 1, not 0. I suspect calling `mkOp(kind, [1,2,3,4])` will throw an exception while `mkOp([kind, 1,2,3,4])` will work fine.

File: src/api/python/cvc5.pxi [hunk 5]
```diff
@@ -1038,70 +1155,84 @@ cdef class Solver:
...
         term.cterm = self.csolver.mkBoolean(val)
         return term
 
     def mkPi(self):
-        """Create a constant representing the number Pi.
+        """
+            Create a constant representing the number Pi.
 
-        :return: a constant representing Pi
+            :return: A constant representing `Pi <cvc5.Kind.Pi>`.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1193]:
```suggestion
            :return: A constant representing :py:obj:`Pi <cvc5.Kind.Pi>`.
```

File: src/api/python/cvc5.pxi [hunk 6]
```diff
@@ -1038,70 +1155,84 @@ cdef class Solver:
...
+        """
+            Create a real constant.
 
-        :param val: the value of the term. Can be an integer, float, or string. It will be formatted as a string before the term is built.
-        :param den: if not None, the value is `val`/`den`
-        :return: a real term with literal value
+            :param val: The value of the term. Can be an integer, float, or
+                        string. It will be formatted as a string before the
+                        term is built.
+            :param den: If not None, the value is `val`/`den`.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1223]:
```suggestion
            :param den: If not None, the value is ``val``/``den``.
```

File: src/api/python/cvc5.pxi [hunk 7]
```diff
@@ -1115,85 +1246,95 @@ cdef class Solver:
         return term
 
     def mkRegexpAll(self):
-        """Create a regular expression all (re.all) term.
+        """
+            Create a regular expression all (re.all) term.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1251]:
```suggestion
            Create a regular expression all (``re.all``) term.
```

File: src/api/python/cvc5.pxi [hunk 8]
```diff
@@ -1115,85 +1246,95 @@ cdef class Solver:
...
+            :return: The all term.
         """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkRegexpAll()
         return term
 
     def mkRegexpAllchar(self):
-        """Create a regular expression allchar (re.allchar) term.
+        """
+            Create a regular expression allchar (re.allchar) term.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1261]:
```suggestion
            Create a regular expression allchar (``re.allchar``) term.
```

File: src/api/python/cvc5.pxi [hunk 9]
```diff
@@ -1115,85 +1246,95 @@ cdef class Solver:
...
+            :return: The allchar term.
         """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkRegexpAllchar()
         return term
 
     def mkRegexpNone(self):
-        """Create a regular expression none (re.none) term.
+        """
+            Create a regular expression none (re.none) term.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1271]:
```suggestion
            Create a regular expression none (``re.none``) term.
```

File: src/api/python/cvc5.pxi [hunk 10]
```diff
@@ -1229,15 +1372,19 @@ cdef class Solver:
...
 
-        - ``Term mkBitVector(int size, int val=0)``
-        - ``Term mkBitVector(int size, string val, int base)``
+            Supports the following arguments:
 
-        :return: a bit-vector literal term
-        :param size: an integer size.
-        :param val: an integer representating the value, in the first form. In the second form, a string representing the value.
-        :param base: the base of the string representation (second form only)
+            - ``Term mkBitVector(int size, int val=0)``
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1380]: The implementation is weird here. Why not simply add `int size` as first argument and get rid of some of the manual checking?

File: src/api/python/cvc5.pxi [hunk 11]
```diff
@@ -1890,21 +2075,23 @@ cdef class Solver:
...
+            SMT-LIB:
 
-        .. code-block:: smtlib
+            .. code-block:: smtlib
 
-            ( define-funs-rec ( <function_decl>^n ) ( <term>^n ) )
+                ( define-funs-rec ( <function_decl>^n ) ( <term>^n ) )
 
-        Create elements of parameter ``funs`` with mkConst().
+            Create elements of parameter ``funs`` with mkConst().
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2088]:
```suggestion
            Create elements of parameter ``funs`` with :py:meth:`mkConst() <cvc5.Solver.mkConst()>`.
```

File: src/api/python/cvc5.pxi [hunk 12]
```diff
@@ -1924,38 +2111,40 @@ cdef class Solver:
...
+            Requires to enable option
+            :ref:`produce-proofs <lbl-option-produce-proofs>`.
+
+            .. warning:: This method is experimental and may change in future
+                         versions.
 
-        :return: a string representing the proof, according to the value of
-                 proof-format-mode.
+            :return: A string representing the proof, according to the value of
+                     proof-format-mode.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2131]:
```suggestion
                     :ref:`proof-format-mode <lbl-option-proof-format-mode>`.
```

File: src/api/python/cvc5.pxi [hunk 13]
```diff
@@ -1984,45 +2174,51 @@ cdef class Solver:
...
+            :param option: The option for which the value is queried.
+            :return: A string representation of the option value.
         """
         return self.csolver.getOption(option.encode()).decode()
 
     def getOptionNames(self):
-        """Get all option names that can be used with `setOption`, `getOption` and `getOptionInfo`.
+        """
+            Get all option names that can be used with
+            :py:meth:`Solver.setOption()`, `:py:meth:`Solver.getOption()`
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2209]:
```suggestion
            :py:meth:`Solver.setOption()`, :py:meth:`Solver.getOption()`
```

File: src/api/python/cvc5.pxi [hunk 14]
```diff
@@ -1984,45 +2174,51 @@ cdef class Solver:
...
+        :return: All option names.
         """
         return [s.decode() for s in self.csolver.getOptionNames()]
 
     def getOptionInfo(self, str option):
-        """Get some information about the given option. Check the `OptionInfo`
-        class for more details on which information is available.
+        """
+            Get some information about the given option. Check the
+            :py:class:`OptionInfo` class for more details on which information
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2219]:
```suggestion
            :py:class:`OptionInfo <cvc5.OptionInfo>` class for more details on which information
```

File: src/api/python/cvc5.pxi [hunk 15]
```diff
@@ -1984,45 +2174,51 @@ cdef class Solver:
...
+            is available.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2220]:
```suggestion
            Get some information about the given option. Returns the information provided by the C++ :cpp:func:`OptionInfo <cvc5::OptionInfo>` as a dictionary.
```

File: src/api/python/cvc5.pxi [hunk 16]
```diff
@@ -2090,8 +2292,11 @@ cdef class Solver:
         return res
 
     def getOptionNames(self):
-       """Get all option names that can be used with `setOption`, `getOption` and `getOptionInfo`.
-       :return: all option names
+       """
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2296]: This method is already defined above. Just remove this variant...

File: src/api/python/cvc5.pxi [hunk 17]
```diff
@@ -2205,122 +2419,130 @@ cdef class Solver:
...
-                :math:'Q x_1...x_k' and
-                :math:'P( x_1...x_i, y_1...y_j )' is a quantifier-free formula
-        :return: a formula :math:'\phi'  such that, given the current set of formulas
-               :math:'A asserted to this solver:
-               - :math:'(A \wedge q)' :math:'(A \wedge \phi)' are equivalent
-               - :math:'\phi' is quantifier-free formula containing only free
-                 variables in :math:'y_1...y_n'.
+            :param q: A quantified formula of the form
+                      :math:`Q\\bar{x_1}... Q\\bar{x}_n. P( x_1...x_i,
+                              y_1...y_j)`
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2455]:
```suggestion
                      :math:`Q\\bar{x_1}... Q\\bar{x}_n. P( x_1 \\dots x_i, y_1 \\dots y_j)`
```

File: src/api/python/cvc5.pxi [hunk 18]
```diff
@@ -2395,245 +2620,259 @@ cdef class Solver:
...
+            <lbl-option-produce-interpolants>` to be set to a mode different
+            from `none`.
 
-        Supports the following variants:
+            Supports the following variants:
 
-        - ``Term getInteprolant(Term conj)``
-        - ``Term getInteprolant(Term conj, Grammar grammar)``
+            - ``Term getInteprolant(Term conj)``
+            - ``Term getInteprolant(Term conj, Grammar grammar)``
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2727]:
```suggestion
            - ``Term getInterpolant(Term conj)``
            - ``Term getInterpolant(Term conj, Grammar grammar)``
```

File: src/api/python/cvc5.pxi [hunk 19]
```diff
@@ -2395,245 +2620,259 @@ cdef class Solver:
...
 
 
     def getInterpolantNext(self):
         """
-        Get the next interpolant. Can only be called immediately after
-        a succesful call to get-interpolant or get-interpolant-next. 
-        Is guaranteed to produce a syntactically different interpolant wrt the
-        last returned interpolant if successful.
+            Get the next interpolant. Can only be called immediately after
+            a succesful call to get-interpolant or get-interpolant-next.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2751]:
```suggestion
            a succesfull call to :py:func:`Solver.getInterpolant()` or :py:func:`Solver.getInterpolantNext()`.
```

File: src/api/python/cvc5.pxi [hunk 20]
```diff
@@ -2395,245 +2620,259 @@ cdef class Solver:
...
+            .. code-block:: smtlib
 
-            ( get-interpolant-next )
+                ( get-interpolant-next )
 
-        Requires to enable incremental mode, and 
-        option :ref:`produce-interpolants <lbl-option-produce-interpolants>` to be set to a mode different from `none`.
+            Requires to enable incremental mode, and option
+            :ref:`produce-interpolants <lbl-option-produce-interpolants>` to be
+            set to a mode different from `none`.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2763]:
```suggestion
            set to a mode different from ``none``.
```

File: src/api/python/cvc5.pxi [hunk 21]
```diff
@@ -2395,245 +2620,259 @@ cdef class Solver:
...
+            Requires to enable option :ref:`produce-abducts
+            <lbl-option-produce-abducts>`.
 
-        Supports the following variants:
+            Supports the following variants:
 
-        - ``Term getAbduct(Term conj)``
-        - ``Term getAbduct(Term conj, Grammar grammar)``
+            - ``Term getAbduct(Term conj)``
+            - ``Term getAbduct(Term conj, Grammar grammar)``
```
Comment (nafur) [inline src/api/python/cvc5.pxi:2792]: Like above, can we use `getAbduct(Term conj, Grammar g=None)`?

File: src/api/python/cvc5.pxi [hunk 22]
```diff
@@ -3219,11 +3471,16 @@ cdef class Statistics:
         }
 
     def __getitem__(self, str name):
-        """Get the statistics information for the statistic called ``name``."""
+        """
```
Comment (nafur) [inline src/api/python/cvc5.pxi:3475]: Ditto (`__getitem__` not document as special member function)

File: src/api/python/cvc5.pxi [hunk 23]
```diff
@@ -3349,242 +3620,284 @@ cdef class Term:
...
         """
-        Asserts :py:meth:`isStringValue()`.
+            Asserts :py:meth:`isStringValue()`.
 
-        .. note::
-           This method is not to be confused with :py:meth:`__str__()` which
-           returns the term in some string representation, whatever data it
-           may hold.
+            .. note::
+            j
```
Comment (nafur) [inline src/api/python/cvc5.pxi:3783]:
```suggestion
```

File: src/api/python/cvc5.pxi [hunk 24]
```diff
@@ -3349,242 +3620,284 @@ cdef class Term:
...
+               This method is not to be confused with :py:meth:`__str__()`
```
Comment (nafur) [inline src/api/python/cvc5.pxi:3784]: `Term.__str__()` is not in the documentation (special member function lalala)

File: src/api/python/cvc5.pxi [hunk 25]
```diff
@@ -3349,242 +3620,284 @@ cdef class Term:
...
-            (union
-                (singleton c1) ... (union (singleton c_{n-1}) (singleton c_n))))
+                (set.union
+                    (set.singleton c1) ...
+                    (set.union (set.singleton c_{n-1}) (set.singleton c_n))))
 
-        where ``c1 ... cn`` are values ordered by id such that
-        ``c1 > ... > cn`` (see also :cpp:func:`cvc5::Term::operator>()`).
+            where :math:`c1 ... cn` are values ordered by id such that
+            :math:`c1 > ... > cn`.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:3886]:
```suggestion
            where :math:`c_1 \dots c_n` are values ordered by id such that
            :math:`c_1 > \cdots > c_n`.
```

File: src/api/python/cvc5.pxi [hunk 26]
```diff
@@ -3636,38 +3956,46 @@ cdef class Term:
...
+            :return: True if the term is a floating-point rounding mode
+                     value.
+        """
         return self.cterm.isRoundingModeValue()
 
     def getRoundingModeValue(self):
         """
-        Asserts isRoundingModeValue().
-        :return: the floating-point rounding mode value held by the term.
+            Asserts isRoundingModeValue().
```
Comment (nafur) [inline src/api/python/cvc5.pxi:3989]:
```suggestion
            Asserts :py:meth:`isRoundingModeValue()`.
```

File: src/api/python/cvc5.pxi (hunk 2 above)
Comment (nafur) [reply src/api/python/cvc5.pxi:577]: Yes, I'd prefer to add this (here and in all other places as well). I'm not so sure about `__str__()`, but maybe we should just show everything that is there?

File: src/api/python/cvc5.pxi (hunk 3 above)
[the remark this answers is not in the document]
Comment (nafur) [reply src/api/python/cvc5.pxi:1084]: Sorry, I was looking at the old code. I'll do a separate PR that gets rid of the decorator.

File: src/api/python/cvc5.pxi (hunk 2 above)
Comment (aniemetz) [reply src/api/python/cvc5.pxi:577]: I will not add this in this PR, but separately.

File: src/api/python/cvc5.pxi (hunk 10 above)
Comment (aniemetz) [reply src/api/python/cvc5.pxi:1380]: This PR is about docs, not the implementation. We should discuss this offline, but will not change any implementation in this PR.

File: src/api/python/cvc5.pxi (hunk 18 above)
Comment (aniemetz) [reply src/api/python/cvc5.pxi:2727]: Ditto as above, postponed.

File: src/api/python/cvc5.pxi (hunk 21 above)
Comment (aniemetz) [reply src/api/python/cvc5.pxi:2792]: Ditto

File: docs/api/python/base/statistics.rst [hunk 27]
```diff
@@ -0,0 +1,6 @@
+Statistics
+==========
+
+.. autoclass:: cvc5.Statistics
+    :members:
+    :undoc-members:
```
Comment (nafur) [inline docs/api/python/base/statistics.rst:6]:
```suggestion
    :special-members:
    :undoc-members:
```

File: docs/api/python/base/statistics.rst [hunk 28]
```diff
@@ -0,0 +1,7 @@
+Statistics
+==========
+
+.. autoclass:: cvc5.Statistics
+    :members:
+    :special-members:
```
Comment (nafur) [inline docs/api/python/base/statistics.rst:6]:
```suggestion
    :special-members: __getitem__
```

File: src/api/python/cvc5.pxi (hunk 2 above)
Comment (nafur) [reply src/api/python/cvc5.pxi:577]: see #8513.

File: src/api/python/cvc5.pxi (hunk 22 above)
Comment (nafur) [reply src/api/python/cvc5.pxi:3485]: see #8513.

**Review by nafur — APPROVED** [review]
LGTM.

**Pull request merged**